Python/ceval.c

```diff
@@ -6750,13 +6757,20 @@ maybe_call_line_trace(Py_tracefunc func, PyObject *obj,
         return -1;
     }
     if (line != -1 && f->f_trace_lines) {
-        /* Trace backward edges (except in 'yield from') or if line number has changed */
-        int trace = line != lastline ||
-            (_PyInterpreterFrame_LASTI(frame) < instr_prev &&
-             // SEND has no quickened forms, so no need to use _PyOpcode_Deopt
-             // here:
-             _Py_OPCODE(*frame->prev_instr) != SEND);
-        if (trace) {
+        int prev_op = _PyOpcode_Deopt[_Py_OPCODE(code[instr_prev])];
+        int next_op = _PyOpcode_Deopt[_Py_OPCODE(*frame->prev_instr)];
+        // Trace before FOR_END, not after, even though a backwards
+        // jump happens after. However, don't trace on the first FOR_END
+        // of the for loop, since we're staying on the same line.
+        // Also don't trace JUMP_NO_INTERRUPT --> SEND.
+        bool line_number_changed = (line != lastline);
+        bool first_iteration = (next_op == FOR_END &&
+                                prev_op == JUMP_FORWARD &&
+                                !line_number_changed);
+        bool for_loop_end = (next_op == FOR_END && !first_iteration);
+        bool back_jump = (_PyInterpreterFrame_LASTI(frame) < instr_prev &&
+                          next_op != SEND && prev_op != FOR_END);
+        if (line_number_changed || for_loop_end || back_jump) {
```

Comment:
I'm not yet totally convinced of the robustness of the `first_iteration` check. It works on the whole test suite, and I couldn't come up with any failing cases, but it feels fragile.

Something like this correctly traces the backwards jumps because the POP_TOP has line number -1:

```python
def for_loop_same_line_with_jump():           # line 0
    for x in (0, 1, 2): +x if x >= 0 else -x  # line 1
```

I don't know of any way to get JUMP_FORWARD targeted at FOR_END unless it's the first loop iteration.

Reply:
I could add something like this if preferred:

```c
#ifdef Py_DEBUG
        if (first_iteration) {
            int prevprev_op = _PyOpcode_Deopt[_Py_OPCODE(code[instr_prev-1])];
            assert(prevprev_op == GET_ITER || prevprev_op == LOAD_FAST);
            if (prevprev_op == LOAD_FAST) {
                assert(_Py_OPARG(code[instr_prev-1]) == 0);
                PyObject *names = frame->f_code->co_localsplusnames;
                PyObject *name = PyTuple_GET_ITEM(names, 0);
                assert(_PyUnicode_EqualToASCIIString(name, ".0"));
            }
        }
#endif
```